### ISSUE TITLE
Relative file resolution (FILE_PWD) not working in configs (env.nu, config.nu)

### ISSUE THREAD

**Issue by texastoland** [nushell/nushell#11465]
### Describe the bug

My Nu config dir is symlinked to work around #893. I'm attempting to `source-env` `default_env.nu` to work around #7371. I'm seeing confusing error messages documented below.

I suspect there's a bug and/or poor error message and/or misunderstanding on my part and/or all of the above ❓

## First Attempt

```nu
# env.nu
source-env default_env.nu
```

### Error

```console
Error: nu::parser::sourced_file_not_found

  × File not found
   ╭─[/Users/texas/Library/Application Support/nushell/env.nu:1:1]
 1 │ source-env default_env.nu
   ·            ───────┬──────
   ·                   ╰── File not found: default_env.nu
   ╰────
  help: sourced files need to be available before your script is run
```

Same error with `./default_env.nu`.

### Context

```console
> '/Users/texas/Library/Application Support/nushell/env.nu'
➜ | path join .. default_env.nu
➜ | ls --long $in
➜ | reject name
╭─────┬────────┬──────────┬────────────┬─────────────┬─────────────┬────────────┬────────┬────────┬──────────┬─────────────┬─────────────┬─────────────╮
│   # │  type  │  target  │  readonly  │    mode     │  num_links  │   inode    │  user  │ group  │   size   │   created   │  accessed   │  modified   │
├─────┼────────┼──────────┼────────────┼─────────────┼─────────────┼────────────┼────────┼────────┼──────────┼─────────────┼─────────────┼─────────────┤
│   0 │ file   │          │ false      │ rw-r--r--   │           1 │  110565660 │ texas  │ staff  │  4.3 KiB │ 2 days ago  │ 2 days ago  │ 2 days ago  │
╰─────┴────────┴──────────┴────────────┴─────────────┴─────────────┴────────────┴────────┴────────┴──────────┴─────────────┴─────────────┴─────────────╯
```

Aside: I was surprised I couldn't pipe to `ls`.

## Second Attempt

Supposing it was related to symlinks I tried to `expand` it.

```nu
# env.nu
source-env (default_env.nu | path expand)
```

### Error

```console
Error:   × Error: nu::shell::not_a_constant
  │ 
  │   × Not a constant.
  │    ╭─[/Users/texas/Library/Application Support/nushell/env.nu:1:1]
  │  1 │ source-env (default_env.nu | path expand)
  │    ·             ───────┬──────
  │    ·                    ╰── Value is not a parse-time constant
  │    ╰────
  │   help: Only a subset of expressions are allowed constants during parsing. Try using the 'const' command or typing the value literally.
  │ 
   ╭─[/Users/texas/Library/Application Support/nushell/env.nu:1:1]
 1 │ source-env (default_env.nu | path expand)
   ·            ───────────────┬──────────────
   ·                           ╰── Encountered error during parse-time evaluation
   ╰────
```

Same error with `./default_env.nu` or assigning it to a `const`.

## Final Attempt

I tried quoting it but it resolved relative to `$env.PWD`. That's entirely reasonable since `source` isn't doing the expansion.

```nu
# env.nu
source-env ('default_env.nu' | path expand)
```

### Error

```console
Error: nu::parser::sourced_file_not_found

  × File not found
   ╭─[/Users/texas/Library/Application Support/nushell/env.nu:1:1]
 1 │ source-env ('default_env.nu' | path expand)
   ·            ────────────────┬───────────────
   ·                            ╰── File not found: /Users/texas/default_env.nu
   ╰────
  help: sourced files need to be available before your script is run
```

Same error with `'./default_env.nu'`.

### How to reproduce

Haven't identified the root cause yet.

### Expected behavior

Expect to load.

### Configuration

| key                | value                                          |
| ------------------ | ---------------------------------------------- |
| version            | 0.88.1                                         |
| branch             |                                                |
| commit_hash        |                                                |
| build_os           | macos-x86_64                                   |
| build_target       | x86_64-apple-darwin                            |
| rust_version       | rustc 1.74.0 (79e9716c9 2023-11-13) (Homebrew) |
| cargo_version      | cargo 1.74.0                                   |
| build_time         | 2023-12-14 17:14:47 +00:00                     |
| build_rust_channel | release                                        |
| allocator          | mimalloc                                       |
| features           | dataframe, default, sqlite, trash, which, zip  |
| installed_plugins  |                                                |

**Comment by fdncred** [comment]
I'm kind of lost because this path will never resolve because env.nu is a file, right?
```nushell
> '/Users/texas/Library/Application Support/nushell/env.nu'
➜ | path join .. default_env.nu
```
This is what it says on my system
```nushell
❯ '/Users/fdncred/Library/Application Support/nushell/scripts/env.nu' | path join .. default_env.nu 
/Users/fdncred/Library/Application Support/nushell/scripts/env.nu/../default_env.nu
```
I'm also wondering if you do something like this, if it will work for you.
```nushell
const var = ('/Users/fdncred/Library/Application Support/nushell/scripts/' | path join default_env.nu)
source $var
```

**Comment by texastoland** [comment]
> ```nushell
> > '/Users/texas/Library/Application Support/nushell/env.nu'
> ➜ | path join .. default_env.nu
> ```

Identical to:

```nushell
> '/Users/texas/Library/Application Support/nushell/env.nu'
::: | path dirname | path join default_env.nu
```

I suppose `path join ..` obfuscates my issue:

1. `source-env default_env.nu` doesn't work despite being in the same directory
2. I don't understand the `not_a_constant` error in my second attempt

> I'm also wondering if you do something like this, if it will work for you.

Both the following work:

```nushell
# where my dotfiles are
source-env ~/.config/nushell/default_env.nu
# symlinked until #7371 is resolved
source-env '~/Library/Application Support/nushell/default_env.nu'
```

I initially also tried to use `$nu.default-config-dir` but it wasn't exposed as a constant either.

Thanks for the help I'm loving it besides the 2 linked issues 🙌🏼

**Comment by fdncred** [comment]
This seems to work fine.
```nushell
const x = ($nu.default-config-dir | path join env.nu)
source-env $x
```

I have no issues sourcing files from $nu.default-config-dir but maybe that may be because my `$env.NU_LIB_DIRS` has that folder in the list, which makes it source-able from anywhere without specifying a path.
```nushell
❯ $env.NU_LIB_DIRS
╭───┬───────────────────────────────────────────────────╮
│ 0 │ C:\Users\username\AppData\Roaming\nushell\scripts │
│ 1 │ D:\nu_scripts                                     │
│ 2 │ C:\Users\username\AppData\Roaming\nushell         │
╰───┴───────────────────────────────────────────────────╯
```
A constant has to be a variable declared with const or a full path to a file. https://www.nushell.sh/book/how_nushell_code_gets_run.html and https://www.nushell.sh/book/thinking_in_nu.html go into how nushell works.

piping into `ls` isn't implemented but a motivated contributor could add it.

**Comment by texastoland** [comment]
> This seems to work fine.

IFL I'm not communicating effectively. Starting from scratch but happy to rewrite the description too.

> I have no issues sourcing files from `$nu.default-config-dir` but maybe that may be because my `$env.NU_LIB_DIRS` has that folder in the list, which makes it source-able from anywhere without specifying a path.

Nothing is loaded yet. Could that be the issue? I want to source the generated config in my user config as described in https://github.com/nushell/nushell/issues/893#issuecomment-1872664324. My config directory happens to be symlinked as described in https://github.com/nushell/nushell/issues/7371#issuecomment-1872665303:

> ```nu
> # env.nu
> source-env default_env.nu
> ```

I expect `default_env.nu` to be resolved relative to `env.nu` as described in the [`0.67` release notes](https://www.nushell.sh/blog/2022-08-16-nushell-0_67.html#relative-paths-in-scripts-are-file-relative-kubouch). Instead the error reports `File not found: default_env.nu` with no indication where it's relative to. I showed above that it exists and is readable. Additionally when piped inside a subexpression it reports `Value is not a parse-time constant` whether prefixed with `./` or not. This may not be a bug but worth noting in the excellent docs.

> This seems to work fine.
> 
> ```nushell
> const x = ($nu.default-config-dir | path join env.nu)
> source-env $x
> ```

Thanks! I don't know what I did wrong before but the following successfully loads:

```nushell
source-env ($nu.default-config-dir | path join default_env.nu)
```

> piping into `ls` isn't implemented but a motivated contributor could add it.

Should I open an issue for backlog? It sounds like a couple lines at most. Rust isn't [1 of my languages](piping) but I look forward to collaboration on nushell/nushell.github.io#1189 🤘🏼

In conclusion is the relative import behavior and constant error expected behavior?

**Comment by texastoland** [comment]
> > I have no issues sourcing files from `$nu.default-config-dir` but maybe that may be because my `$env.NU_LIB_DIRS` has that folder in the list, which makes it source-able from anywhere without specifying a path.
> 
> Nothing is loaded yet. Could that be the issue?

I believe this is 100% the issue!

The function that implements paths relative to `FILE_PWD` as well as `NU_LIB_DIRS` appears to be `find_in_dirs_env`:

https://github.com/nushell/nushell/blob/42bb42a2e1718469133fe3a71ac76c40389bbb30/crates/nu-engine/src/env.rs#L283-L302

It relies on the presence of an environment variable that I only found in 2 places:

1. https://github.com/nushell/nushell/blob/42bb42a2e1718469133fe3a71ac76c40389bbb30/crates/nu-command/src/env/source_env.rs#L66
2. https://github.com/nushell/nushell/blob/42bb42a2e1718469133fe3a71ac76c40389bbb30/crates/nu-cmd-lang/src/core_commands/overlay/use_.rs#L136

But AFAICT loading configs bypasses these pathways:

https://github.com/nushell/nushell/blob/42bb42a2e1718469133fe3a71ac76c40389bbb30/crates/nu-cli/src/config_files.rs#L82-L91

> Rust isn't [1 of my languages](https://github.com/pulls?q=author:texastoland+is:pr)

I don't know Rust but hopefully that's a concrete starting point!

**Comment by texastoland** [comment]
I just confirmed `use utils` can't resolve `$nu.default-config-dir | path join scripts utils.nu) in my ``NU_LIB_DIRS` either.

**Comment by fdncred** [comment]
> I just confirmed use utils in my config.nu can't resolve from my NU_LIB_DIRS either.

I don't get how that's possible. `env.nu` is loaded and parsed first (this is where NU_LIB_DIRS is supposed to be), then config.nu. so NU_LIB_DIRS is available to config.nu. I use it this way all the time.

**Comment by texastoland** [comment]
Can you confirm relative paths work in your `env.nu` and lib paths in `config.nu`? Neither works for me. If so I can at least identify where it's happening 👀

**Comment by fdncred** [comment]
I'm not really sure what you want me to test. I was just saying that this works.
nu_lib_dirs definition in my env.nu
```nushell
$env.NU_LIB_DIRS = [
    ($nu.config-path | path dirname | path join 'scripts')
    '/Users/username/src/nu_scripts'
    ($nu.config-path | path dirname)
]
```
using scripts in my config.nu that only work if nu_lib_dirs is set.
```nushell
# external completions
use "custom-completions/cargo/cargo-completions.nu" *

# git completions and aliases
use custom-completions/git/git-completions.nu *
use aliases/git/git-aliases.nu *

# temperature converter
use "sourced/temp.nu" *

# get_weather
use "modules/weather/get-weather.nu" get_weather
```

**Comment by texastoland** [comment]
I presented an simplified use case to test I don't understand?

> ```nu
> # env.nu
> source-env default_env.nu
> ```

You're right about `NU_LIB_DIRS` I was expecting `use utils` to work instead of `use utils.nu`. My bad 🙏🏼

**Comment by texastoland** [comment]
I reviewed the code again.

> I don't get how that's possible. `env.nu` is loaded and parsed first (this is where NU_LIB_DIRS is supposed to be)

I was wrong `NU_LIB_DIRS` works. But `FILE_PWD` is used for relative paths and (I think) only added to the stack by the `use` and `source commands:

> 1. https://github.com/nushell/nushell/blob/42bb42a2e1718469133fe3a71ac76c40389bbb30/crates/nu-command/src/env/source_env.rs#L66
> 2. https://github.com/nushell/nushell/blob/77f10eb2707a12b0558bac06b43fa932a45fe415/crates/nu-cmd-lang/src/core_commands/use_.rs#L112
> 3. https://github.com/nushell/nushell/blob/42bb42a2e1718469133fe3a71ac76c40389bbb30/crates/nu-cmd-lang/src/core_commands/overlay/use_.rs#L136

But `eval_config_contents` doesn't defer to those commands:

> But AFAICT loading configs bypasses these pathways:
> 
> https://github.com/nushell/nushell/blob/42bb42a2e1718469133fe3a71ac76c40389bbb30/crates/nu-cli/src/config_files.rs#L82-L91

That's why `find_in_dirs_env` isn't resolving the path:

> `find_in_dirs_env` implements paths relative to `FILE_PWD` as well as `NU_LIB_DIRS`:
> 
> https://github.com/nushell/nushell/blob/42bb42a2e1718469133fe3a71ac76c40389bbb30/crates/nu-engine/src/env.rs#L283-L302

**Comment by fdncred** [comment]
Does your example work if you use `source` versus `source-env`. I'm not even sure how well `source-env` is supported these days.

**Comment by texastoland** [comment]
No not `use` either. Based on the above code above I think relative paths were overlooked in `eval_config_contents`.

**Comment by fdncred** [comment]
could've been, that was written/changed 22 months ago so a lot has changed since then

**Comment by texastoland** [comment]
Ping @kubouch from Git blame (I hope that's okay) 🔔 I'm not a Rust person but I'm having an issue with relative file loading and I think (based on clicking around GitHub so maybe wrong) I identified the issue in https://github.com/nushell/nushell/issues/11465#issuecomment-1875444463. Maybe can check eventually 🙏🏼 Anyway really thank you all Nu is incredible!

**Comment by kubouch** [comment]
The parser looks for `currrently_parsed_cwd` field which is set before parsing via [`start_in_file()`](https://github.com/nushell/nushell/blob/1920ece759310ee78e0b1c1c447b8f27adce57b7/crates/nu-protocol/src/engine/engine_state.rs#L286) before starting to parse a file. `$env.FILE_PWD` is intended just for users to use in their scripts. Anyway, it's totally possible we forgot to put `start_in_file()` somewhere.

It's hard to figure out what the issue is because it's missing a minimal working example. There are no clear steps to reproduce the issue. It would be great if you could summarize the reproduction steps, starting from scratch, as simple as possible, in a 1., 2., 3., ... list such that we can quickly reproduce it.

**Comment by texastoland** [comment]
> I presented a simplified use case to test I don't understand? This errors for me OOTB (first line of file).
> 
> > ```nu
> > # env.nu
> > source-env default_env.nu
> > ```

This is the minimal example. I can't source *any* file relative to `env.nu`. Same for `config.nu` *unless* `$nu.default-config-dir` is in `$env.NU_LIB_DIRS`.

**Comment by kubouch** [comment]
Yeah, I can reproduce it. Seems like the start file is not set correctly when evaluating the config files. This should be an easy fix: Add `start_in_file()` calls to the appropriate locations (see my previous comment).

**Comment by texastoland** [comment]
Thanks so much!!! It's a small thing but it makes breaking up my config feel messier.

**Comment by texastoland** [comment]
Closing as duplicate of #7849.
